src/connectionbenchmark/benchmarktask.cpp

```diff
@@ -0,0 +1,119 @@
+/* This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this
+ * file, You can obtain one at http://mozilla.org/MPL/2.0/. */
+
+#include "benchmarktask.h"
+#include "constants.h"
+#include "leakdetector.h"
+#include "logger.h"
+#include "mozillavpn.h"
+#include "networkrequest.h"
+
+#include <QTimer>
+
+namespace {
+Logger logger(LOG_MAIN, "BenchmarkTask");
+}
+
+BenchmarkTask::BenchmarkTask(const BenchmarkType& benchmarkType,
+                             const uint32_t& maxRuntime)
+    : Task("BenchmarkTask"),
+      m_benchmarkType(benchmarkType),
+      m_maxRuntime(maxRuntime) {
+  MVPN_COUNT_CTOR(BenchmarkTask);
+}
+
+BenchmarkTask::~BenchmarkTask() { MVPN_COUNT_DTOR(BenchmarkTask); }
+
+void BenchmarkTask::setState(State state) {
+  logger.debug() << "Set state" << state;
+
+  m_state = state;
+}
+
+void BenchmarkTask::run() {
+  logger.debug() << "Run benchmark task";
+
+  if (m_state == StateCancelled) {
+    emit completed();
+    return;
+  }
+
+  setState(StateActive);
+  m_elapsedTimer.start();
+
+  connect(this, &BenchmarkTask::finished, [this]() {
+    setState(StateInactive);
+    emit completed();
+  });
+
+  if (m_benchmarkType == BenchmarkType::Download) {
+    runBenchmarkDownload();
+  } else if (m_benchmarkType == BenchmarkType::Ping) {
+    runBenchmarkPing();
+  }
+
+  QTimer::singleShot(m_maxRuntime, this, &BenchmarkTask::stop);
+}
+
+void BenchmarkTask::stop() {
+  logger.debug() << "Stop benchmark task";
+
+  if (m_state == StateActive) {
+    if (m_benchmarkType == BenchmarkType::Download) {
+      Q_ASSERT(m_downloadRequest);
+      m_downloadRequest->abort();
+    } else if (m_benchmarkType == BenchmarkType::Ping) {
+      finishedPingBenchmark();
+    }
+
+  } else {
+    setState(StateCancelled);
+  }
+}
+
+void BenchmarkTask::runBenchmarkDownload() {
+  logger.debug() << "Run download benchmark";
+
+  m_downloadRequest =
+      NetworkRequest::createForGetUrl(this, Constants::BENCHMARK_DOWNLOAD_URL);
+
+  connect(m_downloadRequest, &NetworkRequest::requestCompleted, this,
+          [&](const QByteArray& data) {
+            finishedDownloadBenchmark(QNetworkReply::NoError, data);
+          });
+  connect(m_downloadRequest, &NetworkRequest::requestFailed, this,
+          &BenchmarkTask::finishedDownloadBenchmark);
+}
+
+void BenchmarkTask::runBenchmarkPing() {
+  logger.debug() << "Run ping benchmark";
+
+  connect(MozillaVPN::instance()->connectionHealth(),
```

Comment:
Sounds good to me! But keep the benchmarktask and when it runs, expose this value.

Reply:
@oskirby I’m now only taking a single ping, but the task is still using the `pingReceived` signal from `ConnectionHealth` instead of directly returning `MozillaVPN::instance()->connectionHealth()->latency()`. The idea is that in case the speed test is started immediately after turning the VPN on we’re taking at least another sample. Does that make sense or is this unnecessary?

https://github.com/mozilla-mobile/mozilla-vpn-client/pull/2841/files#diff-345be1b07b39c3f4ddeaaa2e98cc6e7147ca1f668c6228a4ad2629160bfdeb8cR28-R37